Add FASTA tileset

101arrowz opened the pull request:
## Description

What was changed in this pull request?
A FASTA tileset was added to Clodius.

Why is it necessary?

Useful for any tracks that need sequence data

## Checklist

-   [ ] Unit tests added or updated
-   [x] Updated CHANGELOG.md
-   [x] Run `black .`

pkerpedjiev commented on clodius/tiles/fasta.py:
```diff
@@ -0,0 +1,260 @@
+
+
+
+def natsort_key(s, _NS_REGEX=re.compile(r"(\d+)", re.U)):
```
This function is defined here: https://github.com/higlass/clodius/blob/develop/clodius/tiles/bigwig.py#L37

Would you mind factoring it out into `clodius.tiles.utils`?

101arrowz replied:
I'll refactor the locations of all of these duplicate functions in the next commit.

pkerpedjiev commented on clodius/tiles/fasta.py:
```diff
@@ -0,0 +1,260 @@
+    return tuple([int(x) if x.isdigit() else x for x in _NS_REGEX.split(s) if x])
+
+
+def natcmp(x, y):
```
Ditto for this function. ☝️

pkerpedjiev commented on clodius/tiles/fasta.py:
```diff
@@ -0,0 +1,260 @@
+                    'max_zoom': 7
+                    }
+    """
+    TILE_SIZE = 1024
```
This is already defined up top.

101arrowz replied:
This was following the precedent from bigwig - will change in both places.

pkerpedjiev commented on clodius/tiles/fasta.py:
```diff
@@ -0,0 +1,260 @@
+
+
+def get_quadtree_depth(chromsizes):
+    tile_size_bp = TILE_SIZE
```
Any reason to do this assignment instead of just using the global in the line below?

101arrowz replied:
This was copied from the other files. I didn't look into why, but it does seem extraneous.

pkerpedjiev commented on clodius/tiles/fasta.py:
```diff
@@ -0,0 +1,260 @@
+        chromsizes_list = chromsizes
+
+    min_tile_cover = np.ceil(sum([int(c[1]) for c in chromsizes_list]) / TILE_SIZE)
+    max_zoom = int(np.ceil(np.log2(min_tile_cover)))
```
You can do `max_zoom = get_quadtree_depth(chromsizes)`, no?

101arrowz replied:
Again, copied from the other files, but I'll make the change.

pkerpedjiev commented on clodius/tiles/fasta.py:
```diff
@@ -0,0 +1,260 @@
+
+    tileset_info = {
+        "min_pos": [0],
+        "max_pos": [TILE_SIZE * 2 ** max_zoom],
```
This is poorly documented, but it should be the sum of the chromsizes.

101arrowz replied:
I thought it was weird too but all the other files seem to use this same system. I based most of this tileset off of bigwig.

101arrowz replied:
Changing this in `bigwig.py` as well (which is where I first found that code block).

pkerpedjiev commented on clodius/tiles/fasta.py:
```diff
@@ -0,0 +1,260 @@
+    return tileset_info
+
+
+def abs2genomic(chromsizes, start_pos, end_pos):
```
OK, we've definitely set a bad precedent by duplicating this function in clodius/tiles/bam.py, clodius.tiles.bigwig.py, and clodius.tiles.multivec.py but if you're feeling generous and want to refactor it into clodius.tiles.utils, it would be greatly appreciated. If not, no worries, we'll do it another time.

101arrowz replied:
Will do.

pkerpedjiev commented on clodius/tiles/fasta.py:
```diff
@@ -0,0 +1,260 @@
+
+def get_chromsizes(fapath):
+    with Fasta(fapath, one_based_attributes=False) as fa:
+        chromsizes = dict((seq, len(fa.records[seq])) for seq in fa.keys())
```
Here you're reading the raw sequence for each chromosome. I'm not too familiar with `pyfaidx` but if you have a `.fai` index of a fasta file, you can read the sizes of each chromosome without reading their actual sequence (see the `samtools faidx` command).

101arrowz replied:
I'm aware that it's not necessary to read the sequence to get the chromsizes, but actually `fa.records[seq]` does not read the sequence; it only reads the index. To read the sequence you must access `fa.records[seq].seq`. This is possible due to clever Python tricks within `pyfaidx`, from my investigation into its source code.

pkerpedjiev reviewed (commented):
Thanks for submitting this! I think it generally looks good. There's one issue that concerns me: when a large fasta file is supplied the `tiles` function could end up returning a ton of data. A solution is to limit the maximum tile width to something sensible like 100K bp.

See here for an example of how we do that with bam files:

https://github.com/higlass/clodius/blob/develop/clodius/tiles/bam.py#L177

Same goes for the `tileset_info` function which AFAICT reads the entire fasta file to determine the sizes of the chromosomes.

101arrowz commented:
I've fixed all of the issues and refactored the duplicate functions. I've also added a `max_tile_width` very similarly to how bam does it. I'd like to have everything working as soon as possible, so I would appreciate it if you could merge this PR now and let me add the unit test in a future PR.

pkerpedjiev reviewed (approved):
Nice! Thank you kindly. Really appreciate the PR!!
